src/ts/ConvivaAnalytics.ts

```diff
@@ -657,79 +698,127 @@ export class ConvivaAnalytics {
     this.convivaVideoAnalytics = null;
+
+    this.convivaAdAnalytics.release();
+    this.convivaAdAnalytics = null;
   };
 
   private onVideoQualityChanged = (event: VideoQualityChangedEvent) => {
+    this.debugLog('[ Player Event ] video quality changed', event);
+
     // We calculate the bitrate with a divisor of 1000 so the values look nicer
     // Example: 250000 / 1000 => 250 kbps (250000 / 1024 => 244kbps)
     const bitrateKbps = Math.round(event.targetQuality.bitrate / 1000);
 
-    if (!this.isSessionActive()) {
-      // Since the first videoPlaybackQualityChanged event comes before playback ever started we need to store the
-      // value and use it for tracking when initializing the session.
-      // TODO: remove this workaround when the player event order is fixed
-      this.lastSeenBitrate = bitrateKbps;
-      return;
-    }
-
-    this.lastSeenBitrate = null;
+    this.debugLog('[ ConvivaAnalytics ] report bitrate', {
+      event,
+      bitrateKbps,
+    });
     this.convivaVideoAnalytics.reportPlaybackMetric(Conviva.Constants.Playback.BITRATE, bitrateKbps);
   };
 
   private onCustomEvent = (event: PlayerEventBase) => {
+    this.debugLog('[ Player Event ] custom playback related event', event);
+
     if (!this.isSessionActive()) {
-      this.debugLog('skip custom event, no session existing', event);
+      this.debugLog('[ ConvivaAnalytics ] skip custom event, no session existing', event);
       return;
     }
 
     const eventAttributes = ObjectUtils.flatten(event);
     this.sendCustomPlaybackEvent(event.type, eventAttributes);
   };
 
-  private trackAdBreakStarted = (event: AdBreakEvent) => {
-    this.debugLog('[ ConvivaAnalytics ] adbreak started', event);
-    this.isAd = true;
+  // Fires before `onAdStarted`
+  private onAdBreakStarted = (event: AdBreakEvent) => {
+    this.debugLog('[ Player Event ] adbreak started', event);
 
-    const adPosition = AdBreakHelper.mapAdPosition(event.adBreak, this.player);
-
-    if (!this.isSessionActive()) {
-      // Don't report without a valid session (e.g., in case of a pre-roll, or post-roll ad)
-      return;
-    }
+    this.isAdBreak = true;
+    this.lastAdBreakEvent = event;
 
+    this.debugLog('[ ConvivaAnalytics ] report ad break started', event);
     this.convivaVideoAnalytics.reportAdBreakStarted(
       Conviva.Constants.AdType.CLIENT_SIDE,
       Conviva.Constants.AdPlayer.SEPARATE,
     );
   };
 
-  private onAdBreakFinished = (event: AdBreakEvent | ErrorEvent) => {
-    this.debugLog('[ ConvivaAnalytics ] adbreak finished', event);
-    this.isAd = false;
+  private onAdStarted = (event: AdEvent) => {
+    this.debugLog('[ Player Event ] ad started', event);
 
-    if (!this.isSessionActive()) {
-      // Don't report without a valid session (e.g., in case of a pre-roll, or post-roll ad)
-      return;
+    const adInfo = AdHelper.extractConvivaAdInfo(this.player, this.lastAdBreakEvent, event);
+    // It's already in kbps
```

Comment:
I think this comment can be removed.

Reply:
Removed in https://github.com/bitmovin/bitmovin-player-web-analytics-conviva/pull/104/commits/763529964e102b9048dfb0aba21f7c1f56163bc1.